From e530dc29b3faa89d3119d20ab0b46d23b152919b Mon Sep 17 00:00:00 2001
From: Vladimir Sokolovsky <vlad@mellanox.co.il>
Date: Thu, 8 Jan 2009 09:45:50 +0200
Subject: [PATCH] mlx4/IB: Add set_4k_mtu module parameter.

It control Infiniband link MTU for all IB ports in a host.

Signed-off-by: Vladimir Sokolovsky <vlad@mellanox.co.il>
---
 drivers/infiniband/hw/mlx4/main.c |    9 +++++++++
 1 files changed, 9 insertions(+), 0 deletions(-)

Index: ofa_1_4_dev_kernel/drivers/net/mlx4/port.c
===================================================================
--- ofa_1_4_dev_kernel.orig/drivers/net/mlx4/port.c
+++ ofa_1_4_dev_kernel/drivers/net/mlx4/port.c
@@ -38,6 +38,10 @@
 
 #include "mlx4.h"
 
+int mlx4_ib_set_4k_mtu = 0;
+module_param_named(set_4k_mtu, mlx4_ib_set_4k_mtu, int, 0444);
+MODULE_PARM_DESC(set_4k_mtu, "attempt to set 4K MTU to all ConnectX ports");
+
 void mlx4_init_mac_table(struct mlx4_dev *dev, struct mlx4_mac_table *table)
 {
 	int i;
@@ -305,8 +309,11 @@ int mlx4_SET_PORT(struct mlx4_dev *dev, 
 		((u8 *) mailbox->buf)[3] = 6;
 		((__be16 *) mailbox->buf)[4] = cpu_to_be16(1 << 15);
 		((__be16 *) mailbox->buf)[6] = cpu_to_be16(1 << 15);
-	} else
+	} else {
+		if (mlx4_ib_set_4k_mtu)
+			((__be32 *) mailbox->buf)[0] |= cpu_to_be32((1 << 22) | (1 << 21) | (5 << 12) | (2 << 4));
 		((__be32 *) mailbox->buf)[1] = dev->caps.ib_port_def_cap[port];
+	}
 	err = mlx4_cmd(dev, mailbox->dma, port, is_eth, MLX4_CMD_SET_PORT,
 		       MLX4_CMD_TIME_CLASS_B);
 
